### PR TITLE
Add `BZPOPMAX` command.

### PR DESCRIPTION
**Pull request by Yury-Fridlyand**
https://redis.io/commands/bzpopmax/

ref: #149 and #174

### CODE REVIEW

File: java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java [hunk 1]
```diff
@@ -272,6 +272,32 @@ CompletableFuture<Double> zaddIncr(
      */
     CompletableFuture<Map<String, Double>> zpopmax(String key);
 
+    /**
+     * Blocks the connection until removes and returns a member with the highest score from the sorted
+     * sets stored at the specified <code>keys</code>.<br>
+     * <code>BZPOPMAX</code> is the blocking variant of {@link #zpopmax(String)}.<br>
+     * A member with the highest score is popped from first sorted set that is non-empty, with the
```
Comment (SanHalacogluImproving) [inline java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java:279]: Do you need this sentence after the first one seems like they are explaining the same thing maybe combine the two sentences?

File: java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java [hunk 2]
```diff
@@ -272,6 +272,32 @@ CompletableFuture<Double> zaddIncr(
...
+     * given <code>keys</code> being checked in the order that they are given.
+     *
+     * @see <a href="https://redis.io/commands/bzpopmax/">redis.io</a> for more details.
+     * @apiNote <code>BZPOPMAX</code> is a client blocking command, see <a
+     *     href="https://github.com/aws/glide-for-redis/wiki/General-Concepts#blocking-commands">Blocking
+     *     Commands</a> for more details and best practices.
+     * @param keys The keys of the sorted sets.
+     * @param timeout The number of seconds to wait for a blocking <code>BZPOPMAX</code> operation to
```
Comment (SanHalacogluImproving) [inline java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java:287]:
```suggestion
     * @param timeout The number of seconds to wait for a blocking operation to
```
Nit: Don't think you need to specify the command name seems trivial.

File: java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java [hunk 3]
```diff
@@ -272,6 +272,32 @@ CompletableFuture<Double> zaddIncr(
...
+     *     complete. A value of <code>0</code> will block indefinitely.
+     * @return An array containing the key where element was popped out (as <code>String</code>), the
```
Comment (SanHalacogluImproving) [inline java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java:289]: Nit: Seems a bit to wordy, an array containing the key name, the popped element and the element  score. (You can keep the types)

File: java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java [hunk 4]
```diff
@@ -272,6 +272,32 @@ CompletableFuture<Double> zaddIncr(
...
+     *     element itself (as <code>String</code>) and the element score (as <code>Double</code>).<br>
+     *     The data could be interpreted in format as <code>[key, element, score]</code>.<br>
```
Comment (SanHalacogluImproving) [inline java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java:291]: Incorprate this sentence to the one above like to be less wordy:
an array containing the key name, the popped element and the element score. [key, element, score]

File: java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java [hunk 5]
```diff
@@ -272,6 +272,32 @@ CompletableFuture<Double> zaddIncr(
...
+     *     If no element could be popped and the timeout expired, returns </code>null</code>.
+     * @example
+     *     <pre>{@code
+     * Object[] data = client.bzpopmax(new String[] {"zset1", "zset2"}, 0.5).get();
+     * System.out.printf("Popped '%s' with score %d from sorted set '%s'%n", data[1], data[2], data[0]);
```
Comment (SanHalacogluImproving) [inline java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java:296]: Do you want to show the explicit cast to the user?

File: java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java [hunk 6]
```diff
@@ -416,7 +442,7 @@ CompletableFuture<Map<String, Double>> zrangeWithScores(
     /**
      * Returns the rank of <code>member</code> in the sorted set stored at <code>key</code>, with
      * scores ordered from low to high.<br>
-     * To get the rank of <code>member</code> with it's score, see <code>zrankWithScore</code>.
+     * To get the rank of <code>member</code> with its score, see {@link #zrankWithScore}.
```
Comment (SanHalacogluImproving) [inline java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java:445]: Other PR?

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 7]
```diff
@@ -1318,7 +1319,7 @@ public T zrem(@NonNull String key, @NonNull String[] members) {
      *     return <code>0</code>.
      */
     public T zcard(@NonNull String key) {
-        ArgsArray commandArgs = buildArgs(new String[] {key});
+        ArgsArray commandArgs = buildArgs(key);
```
Comment (SanHalacogluImproving) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:1322]: Should do this clean up in another PR.

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 8]
```diff
@@ -1390,11 +1391,37 @@ public T zpopmax(@NonNull String key, long count) {
...
      */
     public T zpopmax(@NonNull String key) {
-        ArgsArray commandArgs = buildArgs(new String[] {key});
+        ArgsArray commandArgs = buildArgs(key);
         protobufTransaction.addCommands(buildCommand(ZPopMax, commandArgs));
         return getThis();
     }
 
+    /**
+     * Blocks the connection until removes and returns a member with the highest score from the sorted
```
Comment (SanHalacogluImproving) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:1400]: Update based on comments above.

File: java/integTest/src/test/java/glide/SharedCommandTests.java [hunk 9]
```diff
@@ -1205,6 +1206,32 @@ public void zpopmax(BaseClient client) {
...
+    public void bzpopmax(BaseClient client) {
+        String key1 = "{test}-1-" + UUID.randomUUID();
+        String key2 = "{test}-2-" + UUID.randomUUID();
+        String key3 = "{test}-3-" + UUID.randomUUID();
+
+        assertEquals(2, client.zadd(key1, Map.of("a", 1.0, "b", 1.5)).get());
+        assertEquals(1, client.zadd(key2, Map.of("c", 2.0)).get());
+        assertArrayEquals(
+                new Object[] {key1, "b", 1.5}, client.bzpopmax(new String[] {key1, key2}, .5).get());
+        // nothing popped out
```
Comment (SanHalacogluImproving) [inline java/integTest/src/test/java/glide/SharedCommandTests.java:1221]: You could say Non-Existent Key for the comment

File: java/integTest/src/test/java/glide/SharedCommandTests.java [hunk 10]
```diff
@@ -1205,6 +1206,32 @@ public void zpopmax(BaseClient client) {
...
+    @MethodSource("getClients")
+    public void bzpopmax(BaseClient client) {
+        String key1 = "{test}-1-" + UUID.randomUUID();
+        String key2 = "{test}-2-" + UUID.randomUUID();
+        String key3 = "{test}-3-" + UUID.randomUUID();
+
+        assertEquals(2, client.zadd(key1, Map.of("a", 1.0, "b", 1.5)).get());
+        assertEquals(1, client.zadd(key2, Map.of("c", 2.0)).get());
+        assertArrayEquals(
+                new Object[] {key1, "b", 1.5}, client.bzpopmax(new String[] {key1, key2}, .5).get());
```
Comment (SanHalacogluImproving) [inline java/integTest/src/test/java/glide/SharedCommandTests.java:1220]: Pls add a test where initial set is empty so it checks the second set

File: java/integTest/src/test/java/glide/TransactionTestUtilities.java [hunk 11]
```diff
@@ -98,6 +99,10 @@ public static BaseTransaction<?> transactionTest(BaseTransaction<?> baseTransact
         baseTransaction.zpopmin(key8);
         baseTransaction.zpopmax(key8);
 
+        baseTransaction
+                .zadd(zsetKey2, Map.of("one", 1.0, "two", 2.0, "three", 3.0))
+                .bzpopmax(new String[] {zsetKey2}, .1);
```
Comment (SanHalacogluImproving) [inline java/integTest/src/test/java/glide/TransactionTestUtilities.java:104]: These timeouts won't be an issue right because the set is already full? This could have been an issue if the set was initially empty and would block until it filled up? Just curious thanks!

File: java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java (hunk 5 above)
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java:296]: `%s` and `%d` show it, isn't it?

File: java/integTest/src/test/java/glide/TransactionTestUtilities.java (hunk 11 above)
Comment (Yury-Fridlyand) [reply java/integTest/src/test/java/glide/TransactionTestUtilities.java:104]: Yes, otherwise it blocks the transaction forever. Luckly, we have test timeout to avoid entire CI block.

File: java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java (hunk 6 above)
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java]: Moved to #201

File: java/client/src/main/java/glide/api/models/BaseTransaction.java (hunk 7 above)
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/models/BaseTransaction.java]: #201

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 12]
```diff
@@ -1451,13 +1450,13 @@ public T zscore(@NonNull String key, @NonNull String member) {
...
     public T zrank(@NonNull String key, @NonNull String member) {
-        ArgsArray commandArgs = buildArgs(key, member);
+        ArgsArray commandArgs = buildArgs(new String[] {key, member});
         protobufTransaction.addCommands(buildCommand(Zrank, commandArgs));
         return getThis();
     }
 
     /**
-     * Returns the rank of <code>member</code> in the sorted set stored at <code>key</code> with its
+     * Returns the rank of <code>member</code> in the sorted set stored at <code>key</code> with it's
```
Comment (jonathanl-bq) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:1459]: You had it right the first time. Possessive is "its".

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 13]
```diff
@@ -1433,15 +1432,15 @@ public T bzpopmax(@NonNull String[] keys, double timeout) {
...
+        ArgsArray commandArgs = buildArgs(new String[] {key, member});
         protobufTransaction.addCommands(buildCommand(ZScore, commandArgs));
         return getThis();
     }
 
     /**
      * Returns the rank of <code>member</code> in the sorted set stored at <code>key</code>, with
      * scores ordered from low to high.<br>
-     * To get the rank of <code>member</code> with its score, see {@link #zrankWithScore}.
+     * To get the rank of <code>member</code> with it's score, see <code>zrankWithScore</code>.
```
Comment (jonathanl-bq) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:1443]: its

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 14]
```diff
@@ -1484,6 +1483,7 @@ public T zrankWithScore(@NonNull String key, @NonNull String member) {
      */
     public T pttl(@NonNull String key) {
         ArgsArray commandArgs = buildArgs(key);
+
```
Comment (jonathanl-bq) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:1486]: nit: remove newline to be consistent with other methods here

File: java/integTest/src/test/java/glide/SharedCommandTests.java [hunk 15]
```diff
@@ -1218,11 +1218,14 @@ public void bzpopmax(BaseClient client) {
         assertEquals(1, client.zadd(key2, Map.of("c", 2.0)).get());
         assertArrayEquals(
                 new Object[] {key1, "b", 1.5}, client.bzpopmax(new String[] {key1, key2}, .5).get());
-        // nothing popped out
+        // nothing popped out - key does not exist
```
Comment (jonathanl-bq) [inline java/integTest/src/test/java/glide/SharedCommandTests.java:1221]: nit: newline above this for consistency

File: java/integTest/src/test/java/glide/SharedCommandTests.java [hunk 16]
```diff
@@ -1218,11 +1218,14 @@ public void bzpopmax(BaseClient client) {
...
         assertNull(
                 client
                         .bzpopmax(new String[] {key3}, REDIS_VERSION.isLowerThan("7.0.0") ? 1. : 0.001)
                         .get());
+        // pops from the second key
```
Comment (jonathanl-bq) [inline java/integTest/src/test/java/glide/SharedCommandTests.java:1226]: nit: newline above this for consistency

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 17]
```diff
@@ -1400,20 +1400,19 @@ public T zpopmax(@NonNull String key) {
      * Blocks the connection until removes and returns a member with the highest score from the sorted
      * sets stored at the specified <code>keys</code>.<br>
      * <code>BZPOPMAX</code> is the blocking variant of {@link #zpopmax(String)}.<br>
-     * A member with the highest score is popped from first sorted set that is non-empty, with the
-     * given <code>keys</code> being checked in the order that they are given.
+     * The given <code>keys</code> are being checked in the order that they are given.
```
Comment (jonathanl-bq) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:1403]:
```suggestion
     * The given <code>keys</code> are checked in the order that they are given.
```

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 18]
```diff
@@ -1400,20 +1400,19 @@ public T zpopmax(@NonNull String key) {
...
-     * @return Command Response - An array containing the key where element was popped out (as <code>
-     *     String</code>), the element itself (as <code>String</code>) and the element score (as
-     *     <code>Double</code>).<br>
-     *     The data could be interpreted in format as <code>[key, element, score]</code>.<br>
+     * @param timeout The number of seconds to wait for a blocking operation to complete. A value of
+     *     <code>0</code> will block indefinitely.
+     * @return Command Response - An array containing the key where element was popped out, the
+     *     element itself and the element score.<br>
+     *     The data could be interpreted in format as <code>[String key, String element,
+     *     double score]</code>.<br>
```
Comment (jonathanl-bq) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:1415]: Similar feedback to BZMPOP. This doesn't make sense and probably can be removed or reworded.

File: java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java [hunk 19]
```diff
@@ -442,7 +442,7 @@ CompletableFuture<Map<String, Double>> zrangeWithScores(
     /**
      * Returns the rank of <code>member</code> in the sorted set stored at <code>key</code>, with
      * scores ordered from low to high.<br>
-     * To get the rank of <code>member</code> with its score, see {@link #zrankWithScore}.
+     * To get the rank of <code>member</code> with it's score, see <code>zrankWithScore</code>.
```
Comment (jonathanl-bq) [inline java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java:445]: its

File: java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java [hunk 20]
```diff
@@ -276,19 +276,19 @@ CompletableFuture<Double> zaddIncr(
      * Blocks the connection until removes and returns a member with the highest score from the sorted
      * sets stored at the specified <code>keys</code>.<br>
      * <code>BZPOPMAX</code> is the blocking variant of {@link #zpopmax(String)}.<br>
-     * A member with the highest score is popped from first sorted set that is non-empty, with the
-     * given <code>keys</code> being checked in the order that they are given.
+     * The given <code>keys</code> are being checked in the order that they are given.
```
Comment (jonathanl-bq) [inline java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java:279]:
```suggestion
     * The given <code>keys</code> are checked in the order that they are given.
```

File: java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java [hunk 21]
```diff
@@ -276,19 +276,19 @@ CompletableFuture<Double> zaddIncr(
...
      *     Commands</a> for more details and best practices.
      * @param keys The keys of the sorted sets.
-     * @param timeout The number of seconds to wait for a blocking <code>BZPOPMAX</code> operation to
-     *     complete. A value of <code>0</code> will block indefinitely.
-     * @return An array containing the key where element was popped out (as <code>String</code>), the
-     *     element itself (as <code>String</code>) and the element score (as <code>Double</code>).<br>
-     *     The data could be interpreted in format as <code>[key, element, score]</code>.<br>
+     * @param timeout The number of seconds to wait for a blocking operation to complete. A value of
+     *     <code>0</code> will block indefinitely.
+     * @return An array containing the key where element was popped out, the element itself and the
```
Comment (jonathanl-bq) [inline java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java:288]:
```suggestion
     * @return An array containing the key where the element was popped out, the element itself, and the
```

File: java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java [hunk 22]
```diff
@@ -276,19 +276,19 @@ CompletableFuture<Double> zaddIncr(
      * Blocks the connection until removes and returns a member with the highest score from the sorted
```
Comment (jonathanl-bq) [inline java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java:276]:
```suggestion
     * Blocks the connection until it removes and returns a member with the highest score from the sorted
```

File: java/client/src/main/java/glide/api/models/BaseTransaction.java (hunk 13 above)
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/models/BaseTransaction.java:1443]: I moved this to another PR #201

File: java/client/src/main/java/glide/api/models/BaseTransaction.java (hunk 12 above)
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/models/BaseTransaction.java:1459]: I moved this to another PR #201 and reverted here

File: java/client/src/main/java/glide/api/models/BaseTransaction.java (hunk 14 above)
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/models/BaseTransaction.java:1486]: I moved this to another PR https://github.com/Bit-Quill/glide-for-redis/pull/201 and reverted here

File: java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java (hunk 19 above)
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java:445]: reverted and moved to another PR

File: java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java (hunk 21 above)
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java:288]: member vs element

**Review by SanHalacogluImproving — APPROVED** [review]

**Review by jonathanl-bq — APPROVED** [review]
